test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java

```diff
@@ -437,18 +441,26 @@ public void sendRequest(
                 String action,
                 TransportRequest request,
                 TransportRequestOptions options
-            ) {
-                // don't send anything, the receiving node is unresponsive
-                toClose.add(connection);
+            ) throws IOException {
+                if (connection.isClosed()) {
+                    throw new NodeNotConnectedException(connection.getNode(), "connection already closed");
+                } else if (sendSemaphore.tryAcquire()) {
+                    // don't send anything, the receiving node is unresponsive
+                    toClose.add(connection);
+                    sendSemaphore.release();
+                } else {
+                    connection.sendRequest(requestId, action, request, options);
+                }
             }
 
             @Override
             public void clearCallback() {
                 // close to simulate that tcp-ip eventually times out and closes connection (necessary to ensure transport eventually
                 // responds).
                 try {
+                    sendSemaphore.acquire(Integer.MAX_VALUE);
                     IOUtils.close(toClose);
-                } catch (IOException e) {
+                } catch (IOException | InterruptedException e) {
```

Comment:
nit: on an `InterruptedException` we should restore the thread's interrupted status I think.